src/common/version/src/lib.rs

```diff
@@ -12,24 +12,35 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-const DEFAULT_VALUE: &str = "unknown";
+const UNKNOWN: &str = "unknown";
 
 #[allow(clippy::print_stdout)]
 pub fn setup_git_versions() {
     println!(
         "cargo:rustc-env=GIT_COMMIT={}",
-        build_data::get_git_commit().unwrap_or_else(|_| DEFAULT_VALUE.to_string())
+        build_data::get_git_commit()
+            .as_ref()
+            .map(String::as_str)
+            .unwrap_or(UNKNOWN)
```

Comment:
Perhaps fallback to env var or a meta file generated on release, so that building from source without Git can have proper meta also.

Reply:
> a meta file generated on release

Reply:
> every time when a pr is merged

It sounds like we generate a concomitant commit for each commit. Doesn't like a good idea.

Reply:
How about using the version in `Cargo.toml` instead?

>Perhaps regenerate every time when a pr is merged?

That might be impossible... You don't know your commit hash before committing it 🤣

Reply:
> How about using the version in Cargo.toml instead?

Good idea. The version is important while the branch, commit can be left as unknown. IIRC cargo provide a env var for this.

Reply:
another way is to use [gitoxide](https://github.com/Byron/gitoxide/blob/main/crate-status.md#gix-object), if it worth to add this dependency...

Reply:
* print_version
* short_version
* full_version

Already cover this (in src/cmd). Not a problem.